src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -80,5 +170,60 @@ private static bool IsSessionEnabled(HttpContext context)
         {
             return context.GetFeature<ISessionFeature>() != null;
         }
+
+        internal void EnsureObjectCanBeSerialized(object item)
+        {
+            var itemType = item.GetType();
+            Type[] actualTypes = null;
+
+            if (itemType.IsArray)
+            {
+                itemType = itemType.GetElementType();
+            }
+            else if (itemType.GetTypeInfo().IsGenericType)
+            {
+                if (itemType.ExtractGenericInterface(typeof(IList<>)) != null ||
+                    itemType.ExtractGenericInterface(typeof(IDictionary<,>)) != null)
+                {
+                    actualTypes = itemType.GetGenericArguments();
+                }
+            }
+
+            actualTypes = actualTypes ?? new Type[] { itemType };
+
+            // Throw if the key type of the dictionary is not string.
+            if (actualTypes.Length > 1 && actualTypes[0] != typeof(string))
```

Comment:
Do this early on as part of detecting the `IDictionary<,>`. You can avoid having `Type[] actualTypes` and just have one.